### ISSUE TITLE
Crash in macOS when switching fullscreen window between laptop and external display

### ISSUE THREAD

**Issue by ioan-chera** [fltk/fltk#1022]
**Describe the bug**
FLTK crashes when moving a fullscreen window between displays. I saw it happen when moving between a MacBook Pro screen (with "Retina" display) and an external monitor (1920x1080 and 1:1, normal scaling), in either direction.

**To Reproduce**
1. Have a MacBook (Pro) with scaled ("Retina") display (aren't most of them scaled nowadays?) and a connected external monitor (_not_ scaled)
2. Generate a CMake project, such as to Xcode in this case
3. Build the bundled `fullscreen` sample
4. Make the app window fullscreen
5. Go into "Mission Control"
6. Drag the `fullscreen` space into the other display
7. When entering that display, the view will freeze to the previous display's size and the app will crash, with the debugger triggering to display the following message (see attached):
[crashlog.txt](https://github.com/user-attachments/files/16403763/crashlog.txt)

The key message is:
> Simultaneous accesses to 0x600003ef9f88, but modification requires exclusive access.

It happens every time.

**Expected behavior**
No crash, the window gets resized properly for the new display scale

**Screenshots**
None.

**FLTK Version**
master (tested on c0dba15e4ba9d2e4103a4af8c3862c1a7b8b1769)

**FLTK Configure / Build Options**
Used CMake-GUI to generate Xcode project, no changes in the settings.

**Operating System / Platform:**
 - OS: macOS
 - OS Version: Sonoma 14.5
 - Processor if relevant: Apple M2 Pro

**Comment by ManoloFLTK** [comment]
@ioan-chera  Thanks for this report. Could you, please, post here the backtrace at the crash point after having built the FLTK library in debug mode? This will add a crucial line number information. TIA.

**Comment by ioan-chera** [comment]
I posted the log output in the linked attached `output.txt`  file with the line numbers included (see bullet point 7), do you need more information?

**Comment by ManoloFLTK** [comment]
I've seen file crashlog.txt above. It doesn't contain source line number information
8    fullscreen                         0x0000000104e27470 -[FLView did_view_resolution_change] + 552
only information about what function/method runs. That's because FLTK was not built with -g under Xcode.

**Comment by ioan-chera** [comment]
Ah, you're right, those + numbers aren't real line numbers. I was able to copy-paste the stack trace from the crashing thread though. Notice the source file names followed by line numbers.
```
#0	0x00000001abf8d890 in _swift_runtime_on_report ()
#1	0x00000001abf8dff8 in swift::runtime::AccessSet::insert(swift::runtime::Access*, void*, void*, swift::ExclusivityFlags) ()
#2	0x00000001abf8e05c in swift_beginAccess ()
#3	0x000000019fc0b124 in ___lldb_unnamed_symbol162117 ()
#4	0x000000019fc0b4a8 in ___lldb_unnamed_symbol162127 ()
#5	0x000000019fea59f0 in -[NSViewBackingLayer display] ()
#6	0x00000001a02fc34c in __25-[NSView displayIfNeeded]_block_invoke ()
#7	0x000000019f89a8a4 in NSPerformVisuallyAtomicChange ()
#8	0x000000019f900974 in -[NSView displayIfNeeded] ()
#9	0x0000000104f5f698 in -[FLView did_view_resolution_change] at /Users/ioan/dezvoltare/fltk/src/Fl_cocoa.mm:2388
#10	0x0000000104f5fc28 in -[FLView drawRect:] at /Users/ioan/dezvoltare/fltk/src/Fl_cocoa.mm:2452
#11	0x000000019f9717d4 in _NSViewDrawRect ()
#12	0x00000001a0302cf0 in -[NSView _recursive:displayRectIgnoringOpacity:inContext:stopAtLayerBackedViews:] ()
#13	0x000000019f9711d8 in -[NSView(NSLayerKitGlue) _drawViewBackingLayer:inContext:drawingHandler:] ()
#14	0x000000019fea5f80 in -[NSViewBackingLayer drawInContext:] ()
#15	0x000000019fc0e444 in ___lldb_unnamed_symbol162175 ()
#16	0x000000019fc308ec in ___lldb_unnamed_symbol163256 ()
#17	0x000000019fbde840 in ___lldb_unnamed_symbol160938 ()
#18	0x000000019fc304e8 in ___lldb_unnamed_symbol163255 ()
#19	0x000000019fc2de84 in ___lldb_unnamed_symbol163221 ()
#20	0x000000019fc0e0c8 in ___lldb_unnamed_symbol162173 ()
#21	0x000000019fc0e4a4 in ___lldb_unnamed_symbol162176 ()
#22	0x000000019fea5bb4 in -[NSViewBackingLayer display] ()
#23	0x00000001a4208428 in CA::Layer::display_if_needed(CA::Transaction*) ()
#24	0x00000001a438d1a4 in CA::Context::commit_transaction(CA::Transaction*, double, double*) ()
#25	0x00000001a41eae4c in CA::Transaction::commit() ()
#26	0x000000019bfef130 in __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ ()
#27	0x000000019c0833d8 in ___CFXRegistrationPost_block_invoke ()
#28	0x000000019c083320 in _CFXRegistrationPost ()
#29	0x000000019bfbd678 in _CFXNotificationPost ()
#30	0x000000019d0da4e4 in -[NSNotificationCenter postNotificationName:object:userInfo:] ()
#31	0x00000001a198fa70 in post_coordinated_distributed_notification(CGSNotificationType, void*, unsigned int, void*) ()
#32	0x00000001a16dfc3c in CGSPostLocalNotification ()
#33	0x00000001a16df848 in (anonymous namespace)::notify_datagram_handler(unsigned int, CGSDatagramType, void*, unsigned long, void*) ()
#34	0x00000001a1a7ec0c in CGSDatagramReadStream::dispatchMainQueueDatagrams() ()
#35	0x00000001a1a7eb08 in invocation function for block in CGSDatagramReadStream::mainQueueWakeup() ()
#36	0x000000010574cf2c in _dispatch_call_block_and_release ()
#37	0x000000010574eba4 in _dispatch_client_callout ()
#38	0x0000000105761f08 in _dispatch_main_queue_drain ()
#39	0x0000000105761aac in _dispatch_main_queue_callback_4CF ()
#40	0x000000019c03bad4 in __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ ()
#41	0x000000019bff9258 in __CFRunLoopRun ()
#42	0x000000019bff8434 in CFRunLoopRunSpecific ()
#43	0x00000001a679c19c in RunCurrentEventLoopInMode ()
#44	0x00000001a679bfd8 in ReceiveNextEventCommon ()
#45	0x00000001a679bd30 in _BlockUntilNextEventMatchingListInModeWithFilter ()
#46	0x000000019f857d68 in _DPSNextEvent ()
#47	0x00000001a004d808 in -[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] ()
#48	0x0000000104f5a83c in do_queued_events(double) at /Users/ioan/dezvoltare/fltk/src/Fl_cocoa.mm:832
#49	0x0000000104f5a620 in Fl_Darwin_System_Driver::wait(double) at /Users/ioan/dezvoltare/fltk/src/Fl_cocoa.mm:862
#50	0x0000000104eeea2c in Fl::wait(double) at /Users/ioan/dezvoltare/fltk/src/Fl.cxx:630
#51	0x0000000104eeea68 in Fl::run() at /Users/ioan/dezvoltare/fltk/src/Fl.cxx:650
#52	0x0000000104ee3020 in main at /Users/ioan/dezvoltare/fltk/test/fullscreen.cxx:374
#53	0x000000019bb920e0 in start ()
Enqueued from com.apple.root.default-qos.overcommit (Thread 6) Queue : com.apple.root.default-qos.overcommit (serial)
#0	0x0000000105754940 in dispatch_async ()
#1	0x00000001a16dee78 in CGSSnarfAndDispatchDatagrams ()
#2	0x00000001a1a799a4 in SLSGetNextEventRecordInternal ()
#3	0x00000001a181412c in SLEventCreateNextEvent ()
#4	0x00000001a67a7660 in PullEventsFromWindowServerOnConnection(unsigned int, unsigned char, __CFMachPortBoost*) ()
#5	0x00000001a67a75e4 in MessageHandler(__CFMachPort*, void*, long, void*) ()
#6	0x000000019c029f94 in __CFMachPortPerform ()
#7	0x000000019bffaa6c in __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE1_PERFORM_FUNCTION__ ()
#8	0x000000019bffa98c in __CFRunLoopDoSource1 ()
#9	0x000000019bff9350 in __CFRunLoopRun ()
#10	0x000000019bff8434 in CFRunLoopRunSpecific ()
#11	0x000000019f982188 in _NSEventThread ()
#12	0x00000001057f55c0 in _pthread_start ()
#13	0x00000001057ffaa0 in thread_start ()
```

And some more info about the error is this:
```
Current access (a read) started at

#0	0x00000001abf8e05b in swift_beginAccess ()
#1	0x000000019fc0b124 in ___lldb_unnamed_symbol162117 ()
#2	0x000000019fc0b4a8 in ___lldb_unnamed_symbol162127 ()
#3	0x000000019fea59f0 in -[NSViewBackingLayer display] ()
#4	0x00000001a02fc34c in __25-[NSView displayIfNeeded]_block_invoke ()
#5	0x000000019f89a8a4 in NSPerformVisuallyAtomicChange ()
#6	0x000000019f900974 in -[NSView displayIfNeeded] ()
#7	0x0000000104f5f698 in -[FLView did_view_resolution_change] at /Users/ioan/dezvoltare/fltk/src/Fl_cocoa.mm:2388
#8	0x0000000104f5fc28 in -[FLView drawRect:] at /Users/ioan/dezvoltare/fltk/src/Fl_cocoa.mm:2452
#9	0x000000019f9717d4 in _NSViewDrawRect ()
#10	0x00000001a0302cf0 in -[NSView _recursive:displayRectIgnoringOpacity:inContext:stopAtLayerBackedViews:] ()
#11	0x000000019f9711d8 in -[NSView(NSLayerKitGlue) _drawViewBackingLayer:inContext:drawingHandler:] ()
#12	0x000000019fea5f80 in -[NSViewBackingLayer drawInContext:] ()
#13	0x000000019fc0e444 in ___lldb_unnamed_symbol162175 ()
#14	0x000000019fc308ec in ___lldb_unnamed_symbol163256 ()
#15	0x000000019fbde840 in ___lldb_unnamed_symbol160938 ()
#16	0x000000019fc304e8 in ___lldb_unnamed_symbol163255 ()
#17	0x000000019fc2de84 in ___lldb_unnamed_symbol163221 ()
#18	0x000000019fc0e0c8 in ___lldb_unnamed_symbol162173 ()
#19	0x000000019fc0e4a4 in ___lldb_unnamed_symbol162176 ()
#20	0x000000019fea5bb4 in -[NSViewBackingLayer display] ()
#21	0x00000001a4208428 in CA::Layer::display_if_needed(CA::Transaction*) ()
#22	0x00000001a438d1a4 in CA::Context::commit_transaction(CA::Transaction*, double, double*) ()
#23	0x00000001a41eae4c in CA::Transaction::commit() ()
#24	0x000000019bfef130 in __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ ()
#25	0x000000019c0833d8 in ___CFXRegistrationPost_block_invoke ()
#26	0x000000019c083320 in _CFXRegistrationPost ()
#27	0x000000019bfbd678 in _CFXNotificationPost ()
#28	0x000000019d0da4e4 in -[NSNotificationCenter postNotificationName:object:userInfo:] ()
#29	0x00000001a198fa70 in post_coordinated_distributed_notification(CGSNotificationType, void*, unsigned int, void*) ()
#30	0x00000001a16dfc3c in CGSPostLocalNotification ()
#31	0x00000001a16df848 in (anonymous namespace)::notify_datagram_handler(unsigned int, CGSDatagramType, void*, unsigned long, void*) ()
#32	0x00000001a1a7ec0c in CGSDatagramReadStream::dispatchMainQueueDatagrams() ()
#33	0x00000001a1a7eb08 in invocation function for block in CGSDatagramReadStream::mainQueueWakeup() ()
#34	0x000000010574cf2c in _dispatch_call_block_and_release ()
#35	0x000000010574eba4 in _dispatch_client_callout ()
#36	0x0000000105761f08 in _dispatch_main_queue_drain ()
#37	0x0000000105761aac in _dispatch_main_queue_callback_4CF ()
#38	0x000000019c03bad4 in __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ ()
#39	0x000000019bff9258 in __CFRunLoopRun ()
#40	0x000000019bff8434 in CFRunLoopRunSpecific ()
#41	0x00000001a679c19c in RunCurrentEventLoopInMode ()
#42	0x00000001a679bfd8 in ReceiveNextEventCommon ()
#43	0x00000001a679bd30 in _BlockUntilNextEventMatchingListInModeWithFilter ()
#44	0x000000019f857d68 in _DPSNextEvent ()
#45	0x00000001a004d808 in -[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] ()
#46	0x0000000104f5a83c in do_queued_events(double) at /Users/ioan/dezvoltare/fltk/src/Fl_cocoa.mm:832
#47	0x0000000104f5a620 in Fl_Darwin_System_Driver::wait(double) at /Users/ioan/dezvoltare/fltk/src/Fl_cocoa.mm:862
#48	0x0000000104eeea2c in Fl::wait(double) at /Users/ioan/dezvoltare/fltk/src/Fl.cxx:630
#49	0x0000000104eeea68 in Fl::run() at /Users/ioan/dezvoltare/fltk/src/Fl.cxx:650
#50	0x0000000104ee3020 in main at /Users/ioan/dezvoltare/fltk/test/fullscreen.cxx:374
#51	0x000000019bb920e0 in start ()

Previous access (a modification) started at

#0	0x000000019fc0b23c in ___lldb_unnamed_symbol162119 ()
```

The "current access (read)" assembly looks like:
```
AppKit`-[NSView displayIfNeeded]:
    0x19f9008b4 <+0>:   pacibsp 
    0x19f9008b8 <+4>:   sub    sp, sp, #0x50
    0x19f9008bc <+8>:   stp    x20, x19, [sp, #0x30]
    0x19f9008c0 <+12>:  stp    x29, x30, [sp, #0x40]
    0x19f9008c4 <+16>:  add    x29, sp, #0x40
    0x19f9008c8 <+20>:  mov    x19, x0
    0x19f9008cc <+24>:  adrp   x8, 412099
    0x19f9008d0 <+28>:  add    x8, x8, #0x72             ; sNSViewAllowRecursiveDisplayComputedValue
    0x19f9008d4 <+32>:  ldarb  w8, [x8]
    0x19f9008d8 <+36>:  cmp    w8, #0x2
    0x19f9008dc <+40>:  b.eq   0x19f900910               ; <+92>
    0x19f9008e0 <+44>:  cbnz   w8, 0x19f900924           ; <+112>
    0x19f9008e4 <+48>:  adrp   x0, 418272
    0x19f9008e8 <+52>:  add    x0, x0, #0xe28            ; @"NSViewAllowRecursiveDisplay"
    0x19f9008ec <+56>:  adrp   x2, 412099
    0x19f9008f0 <+60>:  add    x2, x2, #0x72             ; sNSViewAllowRecursiveDisplayComputedValue
    0x19f9008f4 <+64>:  adrp   x16, 2565
    0x19f9008f8 <+68>:  add    x16, x16, #0x364          ; NSViewAllowRecursiveDisplayDefaultValueFunction
    0x19f9008fc <+72>:  paciza x16
    0x19f900900 <+76>:  mov    x3, x16
    0x19f900904 <+80>:  mov    w1, #0x1
    0x19f900908 <+84>:  bl     0x19f823180               ; _NSGetBoolAppConfig
    0x19f90090c <+88>:  tbnz   w0, #0x0, 0x19f900924     ; <+112>
    0x19f900910 <+92>:  adrp   x8, 418071
    0x19f900914 <+96>:  ldr    x0, [x8, #0x790]
    0x19f900918 <+100>: bl     0x1a0af9860               ; objc_msgSend$focusView
    0x19f90091c <+104>: cmp    x0, x19
    0x19f900920 <+108>: b.eq   0x19f900974               ; <+192>
    0x19f900924 <+112>: add    x8, sp, #0x8
    0x19f900928 <+116>: adrp   x16, 417663
    0x19f90092c <+120>: ldr    x16, [x16, #0x678]
    0x19f900930 <+124>: mov    x17, x8
    0x19f900934 <+128>: movk   x17, #0x6ae1, lsl #48
    0x19f900938 <+132>: pacda  x16, x17
    0x19f90093c <+136>: str    x16, [sp, #0x8]
    0x19f900940 <+140>: adrp   x9, 3608
    0x19f900944 <+144>: ldr    d0, [x9, #0x618]
    0x19f900948 <+148>: str    d0, [sp, #0x10]
    0x19f90094c <+152>: add    x8, x8, #0x10
    0x19f900950 <+156>: adrp   x16, 2556
    0x19f900954 <+160>: add    x16, x16, #0x294          ; __25-[NSView displayIfNeeded]_block_invoke
    0x19f900958 <+164>: pacia  x16, x8
    0x19f90095c <+168>: adrp   x8, 418089
    0x19f900960 <+172>: add    x8, x8, #0x708            ; __block_descriptor_40_e8_32o_e5_v8?0l
    0x19f900964 <+176>: stp    x16, x8, [sp, #0x18]
    0x19f900968 <+180>: str    x19, [sp, #0x28]
    0x19f90096c <+184>: add    x0, sp, #0x8
    0x19f900970 <+188>: bl     0x19f89a838               ; NSPerformVisuallyAtomicChange
->  0x19f900974 <+192>: ldp    x29, x30, [sp, #0x40]
    0x19f900978 <+196>: ldp    x20, x19, [sp, #0x30]
    0x19f90097c <+200>: add    sp, sp, #0x50
    0x19f900980 <+204>: retab  

```

The "Previous access (modification)" assembly looks like:
```
AppKit`___lldb_unnamed_symbol162119:
    0x19fc0b208 <+0>:  pacibsp 
    0x19fc0b20c <+4>:  stp    x20, x19, [sp, #-0x20]!
    0x19fc0b210 <+8>:  stp    x29, x30, [sp, #0x10]
    0x19fc0b214 <+12>: add    x29, sp, #0x10
    0x19fc0b218 <+16>: mov    x19, x0
    0x19fc0b21c <+20>: adrp   x8, 411291
    0x19fc0b220 <+24>: ldr    x8, [x8, #0xec0]
    0x19fc0b224 <+28>: add    x20, x20, x8
    0x19fc0b228 <+32>: mov    x0, x20
    0x19fc0b22c <+36>: mov    x1, x19
    0x19fc0b230 <+40>: mov    w2, #0x21
    0x19fc0b234 <+44>: mov    x3, #0x0
    0x19fc0b238 <+48>: bl     0x1a065c0b0               ; symbol stub for: swift_beginAccess
->  0x19fc0b23c <+52>: adrp   x16, 0
    0x19fc0b240 <+56>: add    x16, x16, #0x264          ; ___lldb_unnamed_symbol162120
    0x19fc0b244 <+60>: mov    x17, x19
    0x19fc0b248 <+64>: movk   x17, #0xf45, lsl #48
    0x19fc0b24c <+68>: pacia  x16, x17
    0x19fc0b250 <+72>: mov    x0, x16
    0x19fc0b254 <+76>: mov    x1, x20
    0x19fc0b258 <+80>: ldp    x29, x30, [sp, #0x10]
    0x19fc0b25c <+84>: ldp    x20, x19, [sp], #0x20
    0x19fc0b260 <+88>: retab

**Comment by ManoloFLTK** [comment]
I have now a MacBook Pro connected to an external display and try to reproduce the crash.  I go into mission control by a 3-finger gesture slided up. Then, I don't know how to perform operation 6 above "Drag the fullscreen space into the other display". What's the operation/command to do that?

**Comment by ioan-chera** [comment]
Before going into the mission control, click the green system "stop light" button on the FLTK app window (or hover over it and click "Enter Full Screen").
<img width="294" alt="image" src="https://github.com/user-attachments/assets/0a875d7b-abda-4d68-b1f9-017fe3108c89">

Now go into "mission control" (such as by doing the gesture). Now both your displays should be in that mode.

The display which has the FLTK app should have at least 2 thumbnail rectangles on the top margin: one with "Desktop", and another with the FLTK app (and any other thumbnails with other apps you might have moved to fullscreen, or none other).

Move the mouse to the thumbnail with the FLTK app. It's possible the thumbnail will slightly enlarge, to indicate it can accept user input.

Click and drag it to the top of the other display, and release the mouse button there.

Click on the FLTK app thumbnail now placed on that display. It will crash.

**Comment by ManoloFLTK** [comment]
Sorry, we don't succeed to perform that step "Click and drag it to the top of the other display, and release the mouse button there". The FLTK app remains on the display it started at. But we notice that another fullscreen app, VLC, does the same.

**Comment by ioan-chera** [comment]
I don't know if your system is configured differently. In any case, try to check this (sorry for the language, but you can use the layout as a guideline):

1. On the system menu bar, click the Apple menu and go into System Preferences… (System Settings…, or however it's called in English).
2. In the left sidebar, select "Desktop and Dock"
3. Scroll down the main view until you get to a section titled **Mission Control**. On my computer it looks like this (again, sorry for the localization):
![image](https://github.com/user-attachments/assets/9eec61af-6aef-4ec2-a8a6-e5db7325611f)

Check in particular the fourth option. In English it says **Displays have separate Spaces**. On my computer it's enabled and allows each display to have its separate set of fullscreen views. If it's disabled, I suspect that the display would just be a physical extension like on Windows, which would fail the experiment. Enable that option and try to reproduce the bug.

**Comment by ManoloFLTK** [comment]
Great. After activating "Displays have separate Spaces" and logout/login, we can now move the FLTK app and reproduce the crash. Let's see what I can do.

**Comment by ManoloFLTK** [comment]
It should now be fixed with commit 9c7a473. Can you confirm you also see it fixed?

**Comment by ioan-chera** [comment]
Confirmed, it's fixed! Thanks for the rapid response!
